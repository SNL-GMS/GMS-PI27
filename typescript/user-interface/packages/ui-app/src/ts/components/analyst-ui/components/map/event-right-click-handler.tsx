import { determineActionTargetsFromRightClickAndSetActionTargets } from '@gms/common-util';
import {
  selectOpenEventId,
  selectSelectedEventIds,
  useAppDispatch,
  useAppSelector,
  useDeleteEvents,
  useDuplicateEvents,
  useGetQualifiedAndUnqualifiedEventActionTargetIdsFromSelected,
  useRejectEvents,
  useSetEventActionTargets,
  useSetSelectedEventIds
} from '@gms/ui-state';
import type Cesium from 'cesium';
import React from 'react';
import type { CesiumMovementEvent } from 'resium';

import type { CreateEventMenuState } from '~analyst-ui/common/menus/create-event-menu-item';
import { showEventMenu } from '~analyst-ui/common/menus/event-menu';

import { useSetCloseEvent } from '../events/events-util';
import type { EventRow } from '../events/types';
import { dispatchSetEventId, getMapElement, getMousePositionFromCesiumMovement } from './map-utils';
import type { MapRightClickHandler } from './types';

/**
 * Return the right-click handler for an event on the map.
 */
export const useEventRightClickHandler = (
  setEventId: (eventId: string) => void,
  setCreateEventMenuState: (value: CreateEventMenuState) => void
): MapRightClickHandler => {
  const closeEvent = useSetCloseEvent();
  const openEventId = useAppSelector(selectOpenEventId);
  const dispatch = useAppDispatch();

  const selectedEventIds = useAppSelector(selectSelectedEventIds);

  const setSelectedEventIds = useSetSelectedEventIds();
  const duplicateEvents = useDuplicateEvents();
  const rejectEvents = useRejectEvents();
  const deleteEvents = useDeleteEvents();
  const getQualifiedAndUnqualifiedEventActionTargetIdsFromSelected =
    useGetQualifiedAndUnqualifiedEventActionTargetIdsFromSelected();
  const setEventActionTargets = useSetEventActionTargets();

  return React.useCallback(
    (movement: CesiumMovementEvent, target: Cesium.Entity, latitude: number, longitude: number) => {
      const menuPosition = getMousePositionFromCesiumMovement(movement);
      const entityProperties = target?.properties?.event?.getValue() as EventRow;

      const { actionTargets: eventIdsForAction } =
        determineActionTargetsFromRightClickAndSetActionTargets(
          selectedEventIds,
          entityProperties.id,
          setEventActionTargets
        );

      const eventActionTargetIdsFromSelected =
        getQualifiedAndUnqualifiedEventActionTargetIdsFromSelected();

      showEventMenu(
        new MouseEvent('contextmenu', { clientX: menuPosition.x, clientY: menuPosition.y }),
        {
          latitude,
          longitude,
          selectedEventId: target.properties.id.getValue(),
          isOpen: target.properties.id.getValue() === openEventId,
          entityProperties,
          includeEventDetailsMenuItem: true,
          isMapContextMenu: true,
          openCallback: eventId => {
            dispatchSetEventId(eventId, dispatch, setEventId);
          },
          closeCallback: closeEvent,
          duplicateCallback: () => {
            duplicateEvents(eventIdsForAction);
          },
          rejectCallback: () => {
            rejectEvents(eventIdsForAction);
            // unqualified action targets remain selected after reject
            setSelectedEventIds(eventActionTargetIdsFromSelected.unqualified);
          },
          deleteCallback: () => {
            // filter out the event id's which are qualified action targets
            deleteEvents(eventIdsForAction);
            // unqualified action targets remain selected after delete
            setSelectedEventIds(eventActionTargetIdsFromSelected.unqualified);
          },
          setEventIdCallback: setEventId,
          setCreateEventMenuState
        },
        {
          activeElementOnClose: getMapElement(),
          onClose: () => {
            setEventActionTargets([]);
          }
        }
      );
    },
    [
      closeEvent,
      deleteEvents,
      dispatch,
      duplicateEvents,
      getQualifiedAndUnqualifiedEventActionTargetIdsFromSelected,
      openEventId,
      rejectEvents,
      selectedEventIds,
      setCreateEventMenuState,
      setEventActionTargets,
      setEventId,
      setSelectedEventIds
    ]
  );
};
