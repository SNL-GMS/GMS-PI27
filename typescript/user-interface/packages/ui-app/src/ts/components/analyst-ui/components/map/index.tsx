import { IanDisplays, toDisplayTitle } from '@gms/common-model/lib/displays/types';
import { nonIdealStateWithSpinner } from '@gms/ui-core-components';
import React from 'react';

import type { MapComponentProps } from './map-component';

const IANMapLazy = React.lazy(async () =>
  import(/* webpackChunkName: 'ui-app-ian-map' */ './map-component').then(module => ({
    default: module.IANMap
  }))
);

export function IANMap(props: MapComponentProps) {
  return (
    <React.Suspense
      fallback={nonIdealStateWithSpinner(toDisplayTitle(IanDisplays.MAP), 'Loading Display')}
    >
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <IANMapLazy {...props} />
    </React.Suspense>
  );
}
