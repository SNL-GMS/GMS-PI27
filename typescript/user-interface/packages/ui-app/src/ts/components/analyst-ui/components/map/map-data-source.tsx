import type * as Cesium from 'cesium';
import React from 'react';
import type { CesiumMovementEvent, EventTarget } from 'resium';
import { CustomDataSource } from 'resium';

import { mapIanEntitiesToEntityComponent } from '~analyst-ui/components/map/map-utils';

/**
 * entities to be converted into a custom ian-map-data-source, such that each entity contains the
 * provided handlers. The provided name will be assigned to the resulting datasource.
 */
export interface MapDataSourceProps {
  key: string;
  entities: Cesium.Entity[];
  leftClickHandler?: (targetEntity: Cesium.Entity) => () => void;
  rightClickHandler?: (movement: CesiumMovementEvent, target: EventTarget) => void;
  doubleClickHandler?: (movement: CesiumMovementEvent, target: EventTarget) => void;
  mouseEnterHandler?: (movement: CesiumMovementEvent, target: EventTarget) => void;
  mouseLeaveHandler?: (movement: CesiumMovementEvent, target: EventTarget) => void;
  name: string;
  onMount?: () => void;
  show: boolean;
}

/**
 * Creates a CustomDataSource to add to the cesium map by converting an array of entities into entity components
 * and spreading them into the DataSource
 */
export function MapDataSource(props: MapDataSourceProps): JSX.Element {
  const {
    entities,
    leftClickHandler,
    rightClickHandler,
    doubleClickHandler,
    mouseEnterHandler,
    mouseLeaveHandler,
    name,
    onMount,
    show
  } = props;
  const entityComponents = mapIanEntitiesToEntityComponent(
    entities,
    leftClickHandler,
    rightClickHandler,
    doubleClickHandler,
    mouseEnterHandler,
    mouseLeaveHandler,
    onMount
  );
  return (
    <CustomDataSource name={name} show={show}>
      {...entityComponents}
    </CustomDataSource>
  );
}
