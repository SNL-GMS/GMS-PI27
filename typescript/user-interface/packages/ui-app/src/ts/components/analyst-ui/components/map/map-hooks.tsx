import { EventTypes } from '@gms/common-model';
import { SignalDetectionStatus } from '@gms/common-model/lib/signal-detection';
import { determineActionTargetsFromRightClickAndSetActionTargets } from '@gms/common-util';
import type { AnalystWorkspaceTypes } from '@gms/ui-state';
import {
  analystActions,
  mapActions,
  selectActionTargetEventIds,
  selectActionType,
  selectIsSyncedWithWaveformZoom,
  selectMapLayerVisibility,
  selectOpenEventId,
  selectOpenGoldenLayoutDisplays,
  selectOpenIntervalName,
  selectSelectedEventIds,
  selectSelectedSdIds,
  selectSelectedStationsAndChannelIds,
  selectWorkflowTimeRange,
  setSelectedStationIds,
  useAppDispatch,
  useAppSelector,
  useAssociateSignalDetections,
  useEventStatusQuery,
  useGetEvents,
  useSetSignalDetectionActionTargets,
  useStationsVisibility,
  useUnassociateSignalDetections
} from '@gms/ui-state';
import { selectEventAssociationConflictIds } from '@gms/ui-state/lib/app/state/events/selectors';
import type Cesium from 'cesium';
import React from 'react';
import type { CesiumMovementEvent, EventTarget } from 'resium';

import type { CreateEventMenuState } from '~analyst-ui/common/menus/create-event-menu-item';
import { showSignalDetectionMenu } from '~analyst-ui/common/menus/signal-detection-menu';

import { useSetCloseEvent } from '../events/events-util';
import {
  buildMapEventSource,
  getEventOnDoubleClickHandlers,
  getMapElement,
  getMousePositionFromCesiumMovement,
  intervalIsSelected,
  updateMapEventSources,
  waveformDisplayIsOpen
} from './map-utils';
import { showStationMenu } from './station-menu';
import type { MapEventSource } from './types';

/**
 * The hide/show station context menu should not be available (currently) unless the following is true:
 *
 * 1: An interval is selected
 *
 * 2: The waveform display is open (this condition may change later)
 *
 * if both of these conditions are true, then canOpenContext menu is set to true, and this function also returns that
 */
export const useHideShowContextMenuState = (): boolean => {
  const [canOpenContextMenu, setCanOpenContextMenu] = React.useState(false);
  const openDisplays = useAppSelector(selectOpenGoldenLayoutDisplays);
  const currentInterval = useAppSelector(selectWorkflowTimeRange);
  //! useEffect updates local state
  React.useEffect(() => {
    if (waveformDisplayIsOpen(openDisplays) && intervalIsSelected(currentInterval)) {
      setCanOpenContextMenu(true);
    } else {
      setCanOpenContextMenu(false);
    }
  }, [currentInterval, openDisplays]);

  return canOpenContextMenu;
};

/**
 * Get the map synced value from redux and return it
 *
 * @returns boolean
 */
export const useIsMapSyncedToWaveformZoom = (): boolean => {
  return useAppSelector(selectIsSyncedWithWaveformZoom);
};

/**
 * Set the map synced value into redux
 *
 * @param isSynced boolean
 * @returns void
 */
export const useSetIsMapSyncedToWaveformZoom = (isSynced: boolean): void => {
  const dispatch = useAppDispatch();
  dispatch(mapActions.setIsMapSyncedWithWaveformZoom(isSynced));
};

/**
 * Uses an array of event sources to produce data for map panel props
 *
 */
export const useMapPreferredEventData = (): MapEventSource[] => {
  const timeRange = useAppSelector(selectWorkflowTimeRange);
  const eventQuery = useGetEvents();
  const openIntervalName = useAppSelector(selectOpenIntervalName);
  const eventStatusQuery = useEventStatusQuery();
  const openEventId = useAppSelector(selectOpenEventId);
  const emptyArrayRef = React.useRef<MapEventSource[]>([]);
  const eventInConflictIds = useAppSelector(selectEventAssociationConflictIds);
  const layerVisibility = useAppSelector(selectMapLayerVisibility);
  const eventActionTargets = useAppSelector(selectActionTargetEventIds);
  const eventActionType = useAppSelector(selectActionType);

  return React.useMemo(() => {
    const eventsData =
      eventQuery.data.map(event => {
        let preferredEventHypothesis =
          EventTypes.findPreferredEventHypothesisByOpenStageOrDefaultStage(event, openIntervalName);
        if (preferredEventHypothesis?.preferredLocationSolution === undefined) {
          // Fall back to the parent hypothesis of the preferred
          preferredEventHypothesis = EventTypes.findEventHypothesisParent(
            event,
            preferredEventHypothesis
          );
        }
        if (preferredEventHypothesis === undefined) {
          return undefined;
        }
        const locationSolution = EventTypes.findPreferredLocationSolution(
          preferredEventHypothesis.id.hypothesisId,
          event.eventHypotheses
        );

        const eventStatus = eventStatusQuery.data?.[event.id];
        const eventInConflict = eventInConflictIds.includes(event.id);
        const eventIsActionTarget = eventActionTargets?.includes(event.id);
        const mapEventSource = buildMapEventSource(
          {
            event,
            eventStatus,
            eventIsOpen: openEventId === event.id,
            eventInConflict,
            eventIsActionTarget
          },
          locationSolution,
          openIntervalName,
          timeRange,
          eventActionType as AnalystWorkspaceTypes.EventActionTypes
        );

        return mapEventSource;
      }) || emptyArrayRef.current;
    // update the geoOverlappingEvents attribute now that we can compare the whole list of events
    const updatedEventsData: MapEventSource[] = updateMapEventSources(eventsData, layerVisibility);

    return updatedEventsData;
  }, [
    eventQuery.data,
    openIntervalName,
    timeRange,
    eventStatusQuery.data,
    openEventId,
    eventInConflictIds,
    layerVisibility,
    eventActionTargets,
    eventActionType
  ]);
};

/**
 * Uses an array of event sources to produce data for map panel props
 *
 */
export const useMapNonPreferredEventData = (): MapEventSource[] => {
  const timeRange = useAppSelector(selectWorkflowTimeRange);
  const eventQuery = useGetEvents();
  const openIntervalName = useAppSelector(selectOpenIntervalName);
  const eventStatusQuery = useEventStatusQuery();
  const openEventId = useAppSelector(selectOpenEventId);
  const emptyArrayRef = React.useRef<MapEventSource[]>([]);
  const eventInConflictIds = useAppSelector(selectEventAssociationConflictIds);
  const layerVisibility = useAppSelector(selectMapLayerVisibility);
  const eventActionTargets = useAppSelector(selectActionTargetEventIds);
  const eventActionType = useAppSelector(selectActionType);

  return React.useMemo(() => {
    const mapEventSources: MapEventSource[] = [];

    eventQuery.data.forEach(event => {
      const preferredEventHypothesis =
        EventTypes.findPreferredEventHypothesisByOpenStageOrDefaultStage(event, openIntervalName);
      const eventStatus = eventStatusQuery.data?.[event.id];
      const eventInConflict = eventInConflictIds?.includes(event.id) ?? false;
      const eventIsActionTarget = eventActionTargets?.includes(event.id);
      preferredEventHypothesis.locationSolutions.forEach(locationSolution => {
        if (locationSolution.id !== preferredEventHypothesis.preferredLocationSolution.id) {
          mapEventSources.push(
            buildMapEventSource(
              {
                event,
                eventStatus,
                eventIsOpen: openEventId === event.id,
                eventInConflict,
                eventIsActionTarget
              },
              locationSolution,
              openIntervalName,
              timeRange,
              eventActionType as AnalystWorkspaceTypes.EventActionTypes
            )
          );
        }
      });
    });
    // update the geoOverlappingEvents attribute now that we can compare the whole list of events
    const updatedMapEventSources: MapEventSource[] = updateMapEventSources(
      mapEventSources,
      layerVisibility
    );
    return updatedMapEventSources.length === 0 ? emptyArrayRef.current : updatedMapEventSources;
  }, [
    eventQuery.data,
    layerVisibility,
    openIntervalName,
    eventStatusQuery.data,
    eventInConflictIds,
    eventActionTargets,
    openEventId,
    timeRange,
    eventActionType
  ]);
};

/**
 * The on-left-click handler for station entities displayed on the map, defined as a custom hook.
 */
export const useStationOnClickHandler = (): ((targetEntity: Cesium.Entity) => () => void) => {
  const selectedStations = useAppSelector(selectSelectedStationsAndChannelIds);
  const dispatch = useAppDispatch();
  return React.useCallback(
    (targetEntity: Cesium.Entity) => () => {
      if (targetEntity?.properties?.type?.getValue() === 'Station') {
        if (selectedStations.includes(targetEntity.id) && selectedStations.length === 1) {
          dispatch(setSelectedStationIds([]));
        } else {
          dispatch(setSelectedStationIds([targetEntity.id]));
        }
      }
    },
    [selectedStations, dispatch]
  );
};

/**
 * The on-left-click handler for event entities displayed on the map, defined as a custom hook.
 */
export const useEventOnClickHandler = (): ((targetEntity: Cesium.Entity) => () => void) => {
  const dispatch = useAppDispatch();
  const selectedEvents = useAppSelector(selectSelectedEventIds);
  return React.useCallback(
    (targetEntity: Cesium.Entity) => () => {
      if (targetEntity?.properties?.type?.getValue() === 'Event location') {
        if (selectedEvents.includes(targetEntity.id) && selectedEvents.length === 1) {
          dispatch(analystActions.setSelectedEventIds([]));
        } else {
          dispatch(analystActions.setSelectedEventIds([targetEntity.id]));
        }
      }
    },
    [dispatch, selectedEvents]
  );
};
// */

/**
 * Returns the left-click handler for signal detections on the map display
 *
 * @param
 * @param targetEntity
 */
export const useSdOnClickHandler = (): ((targetEntity: Cesium.Entity) => () => void) => {
  const dispatch = useAppDispatch();
  const selectedSdIds = useAppSelector(selectSelectedSdIds);
  return React.useCallback(
    (targetEntity: Cesium.Entity) => () => {
      if (targetEntity?.properties?.type?.getValue() === 'Signal detection') {
        if (selectedSdIds.includes(targetEntity.id) && selectedSdIds.length === 1) {
          dispatch(analystActions.setSelectedSdIds([]));
        } else {
          dispatch(analystActions.setSelectedSdIds([targetEntity.id]));
        }
      }
    },
    [dispatch, selectedSdIds]
  );
};

/**
 * @return the signal detection double-click handler, which will associate or unassociate
 * the SD that has been double-clicked.
 */
export const useSdOnDoubleClickHandler = () => {
  const unassociateSignalDetections = useUnassociateSignalDetections();
  const associateSignalDetections = useAssociateSignalDetections();
  return React.useCallback(
    (movement: CesiumMovementEvent, targetEntity: EventTarget) => {
      return !targetEntity?.id.properties?.associated?.getValue() ||
        targetEntity?.id.properties?.status?.getValue() === SignalDetectionStatus.OTHER_ASSOCIATED
        ? associateSignalDetections([targetEntity.id.id])
        : unassociateSignalDetections([targetEntity.id.id]);
    },
    [associateSignalDetections, unassociateSignalDetections]
  );
};

/**
 * Returns all click handlers related to signal detections
 */
export const useSignalDetectionClickHandlers = () => {
  const sdOnClickHandler = useSdOnClickHandler();
  const sdOnDoubleClickHandler = useSdOnDoubleClickHandler();

  return {
    sdOnClickHandler,
    sdOnDoubleClickHandler
  };
};

/**
 * Returns the right-click handler for signal detections on the map display
 *
 * @param setPhaseMenuVisibilityCb the callbacks for for setting the phase menu visible
 */
export const useSignalDetectionOnRightClickHandler = (
  setPhaseMenuVisibilityCb: (visibility: boolean) => void,
  setCreateEventMenuState: (value: CreateEventMenuState) => void
) => {
  const selectedSdIds = useAppSelector(selectSelectedSdIds);
  const setSignalDetectionActionTargets = useSetSignalDetectionActionTargets();

  return React.useCallback(
    (
      movement: CesiumMovementEvent,
      target: Cesium.Entity,
      latitude: number,
      longitude: number
    ): void => {
      const menuPosition = getMousePositionFromCesiumMovement(movement);

      // set action target(s)
      determineActionTargetsFromRightClickAndSetActionTargets(
        selectedSdIds,
        target.id,
        setSignalDetectionActionTargets
      );

      showSignalDetectionMenu(
        new MouseEvent('contextmenu', {
          clientX: menuPosition.x,
          clientY: menuPosition.y
        }),
        {
          createEventMenuProps: {
            latitude,
            longitude,
            setCreateEventMenuState
          },
          setPhaseMenuVisibilityCb
        },
        {
          activeElementOnClose: getMapElement(),
          onClose: () => {
            setSignalDetectionActionTargets([]);
          }
        }
      );
    },
    [
      selectedSdIds,
      setCreateEventMenuState,
      setPhaseMenuVisibilityCb,
      setSignalDetectionActionTargets
    ]
  );
};

/**
 * returns the onRightClickHandler function used for bringing up a context menu on the map
 */
export const useStationOnRightClickHandler = (
  setCreateEventMenuState: (value: CreateEventMenuState) => void
) => {
  const canShowContextMenu = useHideShowContextMenuState();
  const { setStationVisibility, isStationVisible } = useStationsVisibility();

  return React.useCallback(
    (
      movement: CesiumMovementEvent,
      target: Cesium.Entity,
      latitude: number,
      longitude: number
    ): void => {
      const menuPosition = getMousePositionFromCesiumMovement(movement);
      showStationMenu(
        new MouseEvent('contextmenu', { clientX: menuPosition.x, clientY: menuPosition.y }),
        {
          target,
          canShowContextMenu,
          latitude,
          longitude,
          setCreateEventMenuState,
          setStationVisibility,
          isStationVisible
        },
        { activeElementOnClose: getMapElement() }
      );
    },
    [canShowContextMenu, setCreateEventMenuState, setStationVisibility, isStationVisible]
  );
};

/**
 * Return the double-click handler for an event on the map.
 */
export const useEventDoubleClickHandler = (setEventId: (eventId: string) => void) => {
  const dispatch = useAppDispatch();
  const closeEvent = useSetCloseEvent();
  const openEventId = useAppSelector(selectOpenEventId);

  return React.useMemo(
    () => getEventOnDoubleClickHandlers(dispatch, openEventId, closeEvent, setEventId),
    [dispatch, openEventId, closeEvent, setEventId]
  );
};
