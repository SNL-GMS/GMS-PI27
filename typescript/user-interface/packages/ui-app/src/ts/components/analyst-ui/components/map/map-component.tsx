import { IanDisplays } from '@gms/common-model/lib/displays/types';
import type GoldenLayout from '@gms/golden-layout';
import {
  selectSelectedPhasesForSignalDetectionsCurrentHypotheses,
  selectSelectedSdIds,
  selectValidActionTargetSignalDetectionIds,
  useAllStations,
  useAppSelector,
  useKeyboardShortcutConfigurations,
  useSetSignalDetectionActionTargets,
  useUpdateSignalDetectionPhase
} from '@gms/ui-state';
import React from 'react';

import { CreateEventDialog } from '~analyst-ui/common/dialogs/create-event/create-event-dialog';
import { PhaseSelectorDialog } from '~analyst-ui/common/dialogs/phase-selector/phase-selector-dialog';
import { MapHotkeys } from '~analyst-ui/common/hotkey-configs/map-hotkey-configs';
import type { CreateEventMenuState } from '~analyst-ui/common/menus/create-event-menu-item';
import { BaseDisplay } from '~common-ui/components/base-display';

import { useVisibleSignalDetections } from '../waveform/waveform-hooks';
import {
  useIsMapSyncedToWaveformZoom,
  useMapNonPreferredEventData,
  useMapPreferredEventData
} from './map-hooks';
import { MapPanel } from './map-panel';

export interface MapComponentProps {
  // passed in from golden-layout
  glContainer?: GoldenLayout.Container;
}

/**
 * Map component. Renders a Cesium map and queries for Station Groups
 */
function MapComponent(props: MapComponentProps) {
  const { glContainer } = props;
  const keyboardShortcuts = useKeyboardShortcutConfigurations();
  const stationData = useAllStations();
  const signalDetectionQuery = useVisibleSignalDetections(useIsMapSyncedToWaveformZoom());
  const preferredEventData = useMapPreferredEventData();
  const nonPreferredEventData = useMapNonPreferredEventData();
  const selectedSdIds = useAppSelector(selectSelectedSdIds);

  // CreateEventDialog setup
  const [createEventMenuState, setCreateEventMenuState] = React.useState<CreateEventMenuState>({
    visibility: false
  });

  // PhaseSelectorDialog setup
  const [phaseMenuVisibility, setPhaseMenuVisibility] = React.useState(false);

  const selectedPhases = useAppSelector(selectSelectedPhasesForSignalDetectionsCurrentHypotheses);

  const signalDetectionPhaseUpdate = useUpdateSignalDetectionPhase();

  const validActionTargetSignalDetectionIds = useAppSelector(
    selectValidActionTargetSignalDetectionIds
  );

  const phaseSelectorCallback = React.useCallback(
    async (phases: string[]) => {
      await signalDetectionPhaseUpdate(validActionTargetSignalDetectionIds ?? [], phases[0]);
    },
    [signalDetectionPhaseUpdate, validActionTargetSignalDetectionIds]
  );

  const setSignalDetectionActionTargets = useSetSignalDetectionActionTargets();
  return (
    <BaseDisplay
      glContainer={glContainer}
      tabName={IanDisplays.MAP}
      className="ian-map-gl-container"
      data-cy="ian-map-container"
    >
      <MapHotkeys
        selectedSignalDetectionsIds={selectedSdIds}
        setPhaseMenuVisibility={setPhaseMenuVisibility}
        setCreateEventMenuState={setCreateEventMenuState}
      >
        <MapPanel
          stationsResult={stationData}
          signalDetections={signalDetectionQuery?.data}
          preferredEventsResult={preferredEventData}
          nonPreferredEventsResult={nonPreferredEventData}
          setPhaseMenuVisibilityCb={setPhaseMenuVisibility}
          setCreateEventMenuState={setCreateEventMenuState}
        />
      </MapHotkeys>
      <PhaseSelectorDialog
        isOpen={phaseMenuVisibility}
        title="Set Phase"
        selectedPhases={selectedPhases}
        phaseSelectorCallback={phaseSelectorCallback}
        closeCallback={() => {
          setPhaseMenuVisibility(false);
          setSignalDetectionActionTargets([]);
        }}
        hotkeyCombo={keyboardShortcuts?.hotkeys?.toggleSetPhaseMenu?.combos[0]}
      />
      {/* Placed outside of MapHotkeys to retain focus */}
      <CreateEventDialog
        isOpen={createEventMenuState.visibility}
        onClose={() => {
          setCreateEventMenuState({ visibility: false, latitude: 0, longitude: 0 });
        }}
        latitude={createEventMenuState.latitude}
        longitude={createEventMenuState.longitude}
      />
    </BaseDisplay>
  );
}

export const IANMap = React.memo(MapComponent);
