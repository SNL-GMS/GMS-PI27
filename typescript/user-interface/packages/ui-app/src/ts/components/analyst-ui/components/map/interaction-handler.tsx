import {
  selectSelectedEventIds,
  selectSelectedSdIds,
  selectSelectedStationsAndChannelIds,
  selectSignalDetections,
  useAppDispatch,
  useAppSelector,
  useSetEventActionTargets,
  useSetSignalDetectionActionTargets
} from '@gms/ui-state';
import * as Cesium from 'cesium';
import * as React from 'react';
import { ScreenSpaceEvent, ScreenSpaceEventHandler } from 'resium';

import type { CreateEventMenuState } from '~analyst-ui/common/menus/create-event-menu-item';
import type { MapHandlerProps } from '~common-ui/components/map/types';

import { useEventRightClickHandler } from './event-right-click-handler';
import {
  eventTooltipLabel,
  handleControlClick,
  handleRightClick,
  stationTooltipLabel,
  tooltipHandleAltClick,
  tooltipHandleMouseMove
} from './interaction-utils';
import { useSignalDetectionOnRightClickHandler, useStationOnRightClickHandler } from './map-hooks';
import { setFocusToMap } from './map-utils';

/**
 * This component creates and ScreenSpaceEventHandler along with a ScreenSpaceEvent of type mousemove
 * so that when an entity on the map has been hovered over a tooltip will appear.
 */
export const useInteractionHandler = (
  setPhaseMenuVisibilityCb: (visibility: boolean) => void,
  setEventId: (eventId: string) => void,
  setCreateEventMenuState: (value: CreateEventMenuState) => void
) => {
  // Get entity right-click handlers
  const stationRightClickHandler = useStationOnRightClickHandler(setCreateEventMenuState);
  const sdRightClickHandler = useSignalDetectionOnRightClickHandler(
    setPhaseMenuVisibilityCb,
    setCreateEventMenuState
  );
  const eventRightClickHandler = useEventRightClickHandler(setEventId, setCreateEventMenuState);

  const dispatch = useAppDispatch();
  const signalDetections = useAppSelector(selectSignalDetections);
  const selectedStationsAndChannelIds = useAppSelector(selectSelectedStationsAndChannelIds);
  const selectedEventIds = useAppSelector(selectSelectedEventIds);
  const selectedSignalDetectionIds = useAppSelector(selectSelectedSdIds);

  const setEventActionTargets = useSetEventActionTargets();
  const setSignalDetectionActionTargets = useSetSignalDetectionActionTargets();

  return React.useCallback(
    function InteractionHandler({ viewer }: MapHandlerProps) {
      if (viewer) {
        // check to see if we have a station tooltip entity to work with if not we add it
        if (!viewer.entities.getById('hoverLabelEntity')) {
          viewer.entities.add(stationTooltipLabel);
        }
        // check to see if we have an event tooltip entity to work with if not we add it
        if (!viewer.entities.getById('eventLabelEntity')) {
          viewer.entities.add(eventTooltipLabel);
        }
        return (
          <ScreenSpaceEventHandler key="IMTHandlers">
            <ScreenSpaceEvent
              action={event =>
                handleRightClick(
                  event,
                  viewer,
                  sdRightClickHandler,
                  stationRightClickHandler,
                  eventRightClickHandler,
                  setCreateEventMenuState
                )
              }
              type={Cesium.ScreenSpaceEventType.RIGHT_CLICK}
            />
            <ScreenSpaceEvent
              action={event =>
                tooltipHandleAltClick(
                  event,
                  viewer,
                  Object.values(signalDetections),
                  setEventActionTargets,
                  setSignalDetectionActionTargets
                )
              }
              type={Cesium.ScreenSpaceEventType.LEFT_DOWN}
              modifier={Cesium.KeyboardEventModifier.ALT}
            />
            <ScreenSpaceEvent
              action={event =>
                handleControlClick(
                  event,
                  viewer,
                  dispatch,
                  selectedStationsAndChannelIds,
                  selectedEventIds,
                  selectedSignalDetectionIds
                )
              }
              type={Cesium.ScreenSpaceEventType.LEFT_DOWN}
              modifier={Cesium.KeyboardEventModifier.CTRL}
            />
            <ScreenSpaceEvent
              action={event => tooltipHandleMouseMove(event, viewer)}
              type={Cesium.ScreenSpaceEventType.MOUSE_MOVE}
            />
            <ScreenSpaceEvent
              action={() => setFocusToMap()}
              type={Cesium.ScreenSpaceEventType.WHEEL}
            />
          </ScreenSpaceEventHandler>
        );
      }
      return null;
    },
    [
      dispatch,
      eventRightClickHandler,
      sdRightClickHandler,
      selectedEventIds,
      selectedSignalDetectionIds,
      selectedStationsAndChannelIds,
      setCreateEventMenuState,
      setEventActionTargets,
      setSignalDetectionActionTargets,
      signalDetections,
      stationRightClickHandler
    ]
  );
};
