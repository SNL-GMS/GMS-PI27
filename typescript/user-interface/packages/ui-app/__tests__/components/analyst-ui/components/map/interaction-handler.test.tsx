import { getStore } from '@gms/ui-state';
import { renderHook } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';

import { useInteractionHandler } from '~analyst-ui/components/map/interaction-handler';

import { stationTooltipLabel } from '../../../../../src/ts/components/analyst-ui/components/map/interaction-utils';

const store = getStore();

function TestReduxWrapper({ children }) {
  return <Provider store={store}>{children}</Provider>;
}

describe('InteractionHandler', () => {
  test('can create a mouse move handler', () => {
    const viewer: any = {
      scene: {
        pickPosition: jest.fn(() => {
          'myPosition';
        })
      },
      entities: {
        getById: jest.fn(() => {
          return undefined;
        }),
        add: jest.fn(() => {
          return stationTooltipLabel;
        })
      }
    };
    const container = renderHook(
      () => {
        const InteractionHandler = useInteractionHandler(jest.fn(), jest.fn, jest.fn);
        return <InteractionHandler viewer={viewer} />;
      },
      { wrapper: TestReduxWrapper }
    );
    expect(container).toMatchSnapshot();
  });
  test('can handle an undefined viewer', () => {
    const viewer: any = undefined;
    const container = renderHook(
      () => {
        const InteractionHandler = useInteractionHandler(jest.fn(), jest.fn, jest.fn);
        return <InteractionHandler viewer={viewer} />;
      },
      { wrapper: TestReduxWrapper }
    );
    expect(container).toMatchSnapshot();
  });
});
