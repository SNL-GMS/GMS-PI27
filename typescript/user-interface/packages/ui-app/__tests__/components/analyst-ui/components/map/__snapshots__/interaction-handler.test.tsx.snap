// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractionHandler can create a mouse move handler 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": <InteractionHandler
      viewer={
        {
          "entities": {
            "add": [MockFunction],
            "getById": [MockFunction],
          },
          "scene": {
            "pickPosition": [MockFunction],
          },
        }
      }
    />,
  },
  "unmount": [Function],
}
`;

exports[`InteractionHandler can handle an undefined viewer 1`] = `
{
  "rerender": [Function],
  "result": {
    "current": <InteractionHandler />,
  },
  "unmount": [Function],
}
`;
